lib/ansible/modules/crypto/openssl_csr.py

```diff
@@ -283,20 +288,25 @@ def __init__(self, module):
         self.request = None
         self.privatekey = None
 
-        self.subject = {
-            'C': module.params['countryName'],
-            'ST': module.params['stateOrProvinceName'],
-            'L': module.params['localityName'],
-            'O': module.params['organizationName'],
-            'OU': module.params['organizationalUnitName'],
-            'CN': module.params['commonName'],
-            'emailAddress': module.params['emailAddress'],
-        }
+        self.subject = [
+            ('C', module.params['countryName']),
+            ('ST', module.params['stateOrProvinceName']),
+            ('L', module.params['localityName']),
+            ('O', module.params['organizationName']),
+            ('OU', module.params['organizationalUnitName']),
+            ('CN', module.params['commonName']),
+            ('emailAddress', module.params['emailAddress']),
+        ]
 
-        if not self.subjectAltName:
-            self.subjectAltName = ['DNS:%s' % self.subject['CN']]
+        if module.params['subject']:
+            self.subject = self.subject + crypto_utils.parse_name_field(module.params['subject'])
+        self.subject = [(entry[0], entry[1]) for entry in self.subject if entry[1]]
 
-        self.subject = dict((k, v) for k, v in self.subject.items() if v)
+        if not self.subjectAltName:
+            for sub in self.subject:
+                if OpenSSL._util.lib.OBJ_txt2nid(to_bytes(sub[0])) == 13:  # 13 is the NID for "commonName"
```

Comment:
Is it accepted that a user specifies various `commonName` field ?

If so, the following task become wrong:

```
- openssl_csr:
    path: /path/to/csr
    privatekey_path: /path/to/pkey
    commonName: www.toto.com
    subject:
      comonName: www.toto2.com
```
What is the supposed output of `subjectAltName` in this case?

  * `DNS:www.toto.com`
  * `DNS:www.toto2.com`
  * `DNS:www.toto.com,DNS:www.toto2.com`

Currently Number 1happens, if this case shouldn't happen maybe a warning should be raised. What do you think?

Reply:
I'm a bit unsure. IMHO I'd just deprecate the fields in the module that are part of `subject`, but I wasn't sure how to best go forward with this, since it would break the current API.

I agree that there could be a warning if there are multiple fields and only one is chosen.

Pinging @cpointner

Reply:
Ok, cleared this one up: Certificates without SANs are deprecated for a LONG time already:

"   If a subjectAltName extension of type dNSName is present, that MUST
   be used as the identity. Otherwise, the (most specific) Common Name
   field in the Subject field of the certificate MUST be used. Although
   the use of the Common Name is existing practice, it is deprecated and
   Certification Authorities are encouraged to use the dNSName instead."

https://tools.ietf.org/html/rfc2818#section-3.1 paragraph 4